src/mlpack/methods/ann/dag_network_impl.hpp

```diff
@@ -0,0 +1,1489 @@
+
+  // equivalent to MultiLayer->Forward(), but with a concat.
+  if (sortedNetwork.size() > 1)
+  {
+    InitializeForwardPassMemory(input.n_cols);
+
+    size_t firstLayer = sortedNetwork.front();
+    network[firstLayer]->Forward(input, layerOutputs.front());
+
+    for (size_t i = 1; i < sortedNetwork.size(); i++)
+    {
+      size_t currentLayer = sortedNetwork[i];
+      const std::vector<size_t>& parents = parentsList[currentLayer];
+
+      // Concatenation
+      if (parents.size() > 1)
+      {
+        size_t axis = layerAxes[currentLayer];
+
+        size_t rows = 1;
+        for (size_t j = 0; j < axis; j++)
+          rows *= network[currentLayer]->InputDimensions()[j];
+
+        size_t slices = input.n_cols;
+        size_t numDims = network[currentLayer]->InputDimensions().size();
+        for (size_t j = axis + 1; j < numDims; j++)
+          slices *= network[currentLayer]->InputDimensions()[j];
+
+        CubeType inputAlias;
+        MakeAlias(inputAlias, layerInputs[i - 1], rows,
+          network[currentLayer]->InputDimensions()[axis], slices);
+        size_t startCol = 0;
+
+        for (size_t j = 0; j < parents.size(); j++)
+        {
+          size_t index = sortedIndices[parents[j]];
+          MatType& parentOutput = layerOutputs[index];
+
+          const size_t cols = network[parents[j]]->OutputDimensions()[axis];
+          CubeType parentOutputAlias;
+          MakeAlias(parentOutputAlias, parentOutput, rows, cols, slices);
+
+          inputAlias.cols(startCol, startCol + cols - 1) = parentOutputAlias;
+          startCol += cols;
+        }
+      }
+
+      // Don't execute if it's the last iteration.
+      if (i < sortedNetwork.size() - 1)
+        network[currentLayer]->Forward(layerInputs[i - 1], layerOutputs[i]);
+    }
+    network[lastLayer]->Forward(layerInputs.back(), networkOutput);
+  }
+  else if (sortedNetwork.size() == 1)
+  {
+    network[0]->Forward(input, networkOutput);
+  }
+  else
+  {
+    throw std::invalid_argument("DAGNetwork::Forward(): Cannot use network"
+        " with no layers!");
+  }
+
+  if (&results != &networkOutput)
+    results = networkOutput;
+}
+
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void DAGNetwork<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::Backward(const MatType& input,
+            const MatType& output,
+            const MatType& error,
+            MatType& gradients)
+{
+  CheckNetwork("DAGNetwork::Backward()", input.n_rows, true, true);
+  gradients.set_size(parameters.n_rows, parameters.n_cols);
+  networkDelta.set_size(input.n_rows, input.n_cols);
+
+  if (sortedNetwork.size() > 1)
+  {
+    InitializeBackwardPassMemory(input.n_cols);
+
+    for (size_t i = 0; i < sortedNetwork.size(); i++)
+    {
+      size_t currentLayer = sortedNetwork[i];
+      if (childrenList[currentLayer].size() > 1)
+        accumulatedDeltas[i]->fill(0.0f);
```

Comment:
```suggestion
        accumulatedDeltas[i]->zeros();
```
This would actually be necessary to work correctly with FP16 types, where a `-Wnarrowing` warning would be thrown here.